File: day02/keypad.pl
#!/usr/bin/env perl
use strict;
use warnings;
use List::Util qw(min max);
use 5.24.0;

my $keypad = [
              [1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]
             ];

my ($x, $y) = (1, 1);
my @code;

while (<>) {
    chomp;
    my @moves = split //;
    for my $move (@moves) {
        $y = max($y-1, 0) if $move eq 'U';
        $y = min($y+1, 2) if $move eq 'D';
        $x = max($x-1, 0) if $move eq 'L';
        $x = min($x+1, 2) if $move eq 'R';
#        say "$move $x $y $keypad->[$y][$x]";
    }
#    say "adding $keypad->[$y][$x]";
    push @code, $keypad->[$y][$x];
}

say join "", @code;
# say scalar @code;
# for my $i (0..$#code) {
#     say "$i $code[$i]";
# }

File: day02/keypad.py
#!/usr/bin/env python
from sys import argv

keypad = [
    [0, 0, 0, 0, 0],
    [0, 1, 2, 3, 0],
    [0, 4, 5, 6, 0],
    [0, 7, 8, 9, 0],
    [0, 0, 0, 0, 0] ]

(x, y) = (2, 2)
code = []

script, filename = argv
with open(filename) as f :
    for line in f :
        line = line.rstrip("\n")
        for move in line :
            if move == 'U' and keypad[y-1][x] != 0 :
                y = y - 1
            elif move == 'D' and keypad[y+1][x] != 0 :
                y = y + 1
            elif move == 'L' and keypad[y][x-1] != 0 :
                x = x - 1
            elif move == 'R' and keypad[y][x+1] != 0:
                x = x + 1

        code.append(keypad[y][x])

print ''.join(map(str,code))
